tools/sim/bridge.py

```diff
@@ -236,36 +246,120 @@ def can_function_runner(vs: VehicleState, exit_event: threading.Event):
+    self.world = world
+    self.vc = carla.VehicleControl(throttle=0, steer=0, brake=0, reverse=False)
+    self.vehicle = vehicle
+    self.max_steer_angle: float = vehicle.get_physics_control().wheels[0].max_steer_angle
+    self.camerad = camerad
+
+  def apply_controls(self, steer_sim, throttle_out, brake_out, _frame):
+    self.vc.throttle = throttle_out / 0.6
+    self.vc.steer = steer_sim
+    self.vc.brake = brake_out
+    self.vehicle.apply_control(self.vc)
+
+  def get_velocity(self):
+    self.vel = self.vehicle.get_velocity()
+    return self.vel
+
+  def get_speed(self) -> float:
+    return math.sqrt(self.vel.x ** 2 + self.vel.y ** 2 + self.vel.z ** 2)  # in m/s
+
+  def get_steer_correction(self) -> float:
+    return self.max_steer_angle * STEER_RATIO * -1  
+  
+  def tick(self):
+    self.world.tick()
+
+class MetaDriveWorld(World):
+  def __init__(self, env, frames_per_tick: float):
+    self.env = env
+    self.speed = 0.0
+    self.yuv = None
+    self.camerad = Camerad()
+    self.frames_per_tick = frames_per_tick
+
+  def apply_controls(self, steer_sim, throttle_out, brake_out, frame: int):
+    vc = [0.0, 0.0]
+    vc[0] = steer_sim * -1
+    if throttle_out:
+      vc[1] = throttle_out * 10
+    else:
+      vc[1] = -brake_out
+
+    if frame % self.frames_per_tick == 0 or self.yuv is None:
+      if frame % (self.frames_per_tick * 2) == 0:
+        o, _, _, _ = self.env.step(vc)
+        self.speed = o["state"][3] * self.frames_per_tick * 2 # empirically derived
+      img = self.env.vehicle.image_sensors["rgb_wide"].get_pixels_array(self.env.vehicle, False)
+      self.yuv = self.camerad.img_to_yuv(img)
+
+    if frame % (self.frames_per_tick * 2) == 0:
+      self.camerad.cam_send_yuv_wide_road(self.yuv)
+
+  def get_velocity(self):
+    return None
 
-def connect_carla_client(host: str, port: int):
-  client = carla.Client(host, port)
-  client.set_timeout(5)
-  return client
+  def get_speed(self) -> float:
+    return self.speed
 
+  def get_steer_correction(self) -> float:
+    max_steer_angle = 75 / self.frames_per_tick
+    return max_steer_angle * STEER_RATIO * -1
+  
+  def tick(self):
+    pass
 
-class CarlaBridge:
 
+class SimulatorBridge(ABC):
+  FRAMES_PER_TICK = 5
   def __init__(self, arguments):
+    if arguments.frames_per_tick:
+      self.frames_per_tick = arguments.frames_per_tick
+    else:
+      self.frames_per_tick = self.FRAMES_PER_TICK
+    print(f"Running at {self.frames_per_tick} frames per tick")
     set_params_enabled()
-
     self.params = Params()
 
     msg = messaging.new_message('liveCalibration')
     msg.liveCalibration.validBlocks = 20
     msg.liveCalibration.rpyCalib = [0.0, 0.0, 0.0]
     self.params.put("CalibrationParams", msg.to_bytes())
+    if not arguments.dual_camera:
+      print("Dual Camera disabled")
     self.params.put_bool("WideCameraOnly", not arguments.dual_camera)
+    self.params.put_bool("DisengageOnAccelerator", True)
```

Comment:
Should we add this...? 

Hmmm... But anw we can't `throttle_out` while openpilot selfdrive is running...